medcat-v2/medcat/components/linking/embedding_linker.py

```diff
@@ -0,0 +1,583 @@
+
+    def _get_context_vectors(
+        self, doc: MutableDocument, entities: list[MutableEntity], size: int
+    ) -> Tensor:
+        """Get context vectors for all detected concepts based on their 
+        surrounding text.
+
+        Args:
+            doc (BaseDocument): The document look in.
+            size (int): The size of the entity.
+        Returns:
+            tuple[list[BaseToken], list[BaseToken], list[BaseToken]]:
+                The tokens on the left, centre, and right."""
+        texts = []
+        for entity in entities:
+            text = self._get_context(entity, doc, size)
+            texts.append(text)
+        return self._embed(texts, self.device)
+
+    def _set_filters(self) -> None:
+        include_set = self.cnf_l.filters.cuis
+        exclude_set = self.cnf_l.filters.cuis_exclude
+
+        # Check if sets changed (avoid recomputation if same)
+        if (
+            self._last_include_set is not None
+            and self._last_exclude_set is not None
+            and include_set == self._last_include_set
+            and exclude_set == self._last_exclude_set
+        ):
+            return
+
+        n = len(self._name_keys)
+        allowed_mask = torch.empty(n, dtype=torch.bool, device=self.device)
+
+        if include_set:
+            # if in include set, ignore exclude set.
+            allowed_mask[:] = False
+            include_cui_idxs = {
+                self._cui_to_idx[cui] for cui in include_set if cui in self._cui_to_idx
+            }
+            include_idxs = [
+                name_idx
+                for name_idx, name_cui_idxs in enumerate(self._name_to_cui_idxs)
+                if any(cui in include_cui_idxs for cui in name_cui_idxs)
+            ]
+            allowed_mask[
+                torch.tensor(include_idxs, dtype=torch.long, device=self.device)
+            ] = True
+        else:
+            # only look at exclude if there's no include set
+            allowed_mask[:] = True
+            if exclude_set:
+                exclude_cui_idxs = {
+                    self._cui_to_idx[cui]
+                    for cui in exclude_set
+                    if cui in self._cui_to_idx
+                }
+                exclude_idxs = [
+                    i
+                    for i, name_cui_idxs in enumerate(self._name_to_cui_idxs)
+                    if any(ci in exclude_cui_idxs for ci in name_cui_idxs)
+                ]
+                allowed_mask[
+                    torch.tensor(exclude_idxs, dtype=torch.long, device=self.device)
+                ] = False
+
+        # checking if a name has at least 1 cui related to it.
+        _has_cuis_all = torch.tensor(
+            [
+                bool(self.cdb.name2info[name]["per_cui_status"])
+                for name in self._name_keys
+            ],
+            device=self.device,
+        )
+        self._valid_names = _has_cuis_all & allowed_mask
+        self._last_include_set = include_set
```

Comment:
You probably want a copy of these. Otherwise changes to the set (i.e adding, removing elements) wouldn't result in a recalculation here because it's the same instance as the "last" ones.
Currently, if someone wrote a fully new filter (either exclude or include), the recalculation would be triggered, but not if someone adds / removes stuff to / from existing sets.